src/DataStructures/Tensor/IndexType.hpp

```diff
@@ -110,7 +110,6 @@ inline std::ostream& operator<<(std::ostream& os, const IndexType& index_type) {
 }
 /// \endcond
 
-namespace Tensor_detail {
```

Comment:
Why? Users are not supposed to interact with this directly, it is intentionally in a detail namespace. Under what circumstances does a user need to know the details?

Reply:
It is mentioned 15 times on the documentation page for the Tensor namespace, which it should not be if no one is supposed to know about it (and also appears in other documentation).  For actual uses, the documented public type alias Tensor::index_list contains them, but you can't get any information from them without knowing the members.  The members are used in PartialDerivatives.hpp and Divergence.hpp.

Reply:
Here are all the documentation uses I found with `get grep -n "TensorIndexType"`. None of these link to the class itself but rather to the type aliases.

```
src/DataStructures/Tensor/IndexType.hpp:18:/// Whether a \ref SpacetimeIndex "TensorIndexType" is covariant or
src/DataStructures/Tensor/IndexType.hpp:34:/// Indicates the ::Frame that a \ref SpacetimeIndex "TensorIndexType"
src/DataStructures/Tensor/IndexType.hpp:98:/// Indicates whether the \ref SpacetimeIndex "TensorIndexType" is
src/DataStructures/Tensor/IndexType.hpp:101:  /// The \ref SpatialIndex "TensorIndexType" is purely spatial
src/DataStructures/Tensor/IndexType.hpp:103:  /// The \ref SpacetimeIndex "TensorIndexType" is a spacetime index
src/DataStructures/Tensor/IndexType.hpp:115:/// A ::TensorIndexType holds information about what type of index an index of a
src/DataStructures/Tensor/IndexType.hpp:146:/// SpatialIndex "TensorIndexType"
src/DataStructures/Tensor/IndexType.hpp:148:/// \tparam Fr the ::Frame the \ref SpatialIndex "TensorIndexType"
src/DataStructures/Tensor/IndexType.hpp:160:/// SpacetimeIndex "TensorIndexType"
src/DataStructures/Tensor/IndexType.hpp:162:/// \tparam Fr the ::Frame the \ref SpacetimeIndex "TensorIndexType"
src/DataStructures/Tensor/IndexType.hpp:172:/// "TensorIndexType"
src/DataStructures/Tensor/IndexType.hpp:189:/// Change the \ref SpacetimeIndex "TensorIndexType" to be covariant
src/DataStructures/Tensor/IndexType.hpp:195:/// \tparam Index the \ref SpacetimeIndex "TensorIndexType" to change
src/DataStructures/Tensor/Tensor.hpp:72: * \tparam IndexList a typelist of \ref SpacetimeIndex "TensorIndexType"'s
src/DataStructures/Tensor/Tensor.hpp:108:  /// Typelist of the \ref SpacetimeIndex "TensorIndexType"'s that the
src/DataStructures/Tensor/Tensor.hpp:395:  /// canonical \ref SpacetimeIndex "TensorIndexType"
```

Did I miss the ones you were referring to? For these, depending on what we decide to do, it might make more sense to change the documentation. I don't think moving TensorIndexType should be based on the above documentation.

I also think that `Tensor::index_list` should be public, even though "you can't get any information from it". That's not true, I can get what each index is, and I can then raise and lower it. I don't need to know about `TensorIndexType` to raise and lower an index. I'm not sure what else you'd want to do with the index list other than split it or raise and lower indices. The rest of the index information is provided via public interfaces in `Tensor`.

The actual question is whether having TensorIndexType be "public" is good. If I recall correctly the reason for not having it be public was to have people use the type aliases for spatial and spacetime indices. This works fine until you get to generic code where you want to match either a spatial or spacetime index, such as computing Christoffel symbols. The other issue is there needs to be some way to access the information directly from the index in cases where a full `Tensor` is not available, such as the derivatives you pointed out. Given how we generally wrap the `Tensor` types in the `tnsr` namespace, I'm not sure how much of an impedement a private `TensorIndexType` is to generic programming outside of Tensor internals (I'm not saying it's not, I just don't know). So that really leaves the accessing members, which could be done via public `constexpr` helper variables, or making the class public.

If `TensorIndexType` becomes public the PR will need to add a discussion on `TensorIndexType` linking to the type aliases, and should probably also point out that in most cases this information should just be retrieved from `Tensor` itself. The commit message may also need to be updated because I don't know where it's not linking correctly.